File: ConvenientFileManagerTests/Tests/Persistence/NSFileManagerCFMPersistenceTests.m
//
//  NSFileManagerCFMPersistenceTests.m
//  ConvenientFileManager
//
//  Created by Home on 27/02/2016.
//  Copyright © 2016 Boles. All rights reserved.
//

#import <XCTest/XCTest.h>

#import "NSFileManager+CFMPersistence.h"
#import "NSFileManager+CFMDocuments.h"
#import "NSFileManager+CFMCache.h"

@interface NSFileManagerCFMPersistenceTests : XCTestCase

@property (nonatomic, strong) NSData *dataToBeSaved;
@property (nonatomic, strong) NSString *absolutePath;
@property (nonatomic, strong) NSString *absolutePathWithAdditionalDirectory;
@property (nonatomic, strong) NSString *absolutePathWithoutFile;
@property (nonatomic, strong) NSString *absoluteSourcePath;
@property (nonatomic, strong) NSString *absoluteDestinationPath;

@end

@implementation NSFileManagerCFMPersistenceTests

#pragma mark - TestSuiteLifecycle

- (void)setUp
{
    [super setUp];
    
    self.dataToBeSaved = [@"Test string to be converted into data" dataUsingEncoding:NSUTF8StringEncoding];
    self.absolutePath = [NSFileManager cfm_documentsDirectoryPathForResourceWithPath:@"test.mp4"];
    self.absolutePathWithAdditionalDirectory = [NSFileManager cfm_cacheDirectoryPathForResourceWithPath:@"test/test/test.mp4"];
    self.absolutePathWithoutFile = [NSFileManager cfm_cacheDirectoryPathForResourceWithPath:@"test/testcache"];
    self.absoluteSourcePath = [NSFileManager cfm_cacheDirectoryPathForResourceWithPath:@"test/source"];
    self.absoluteDestinationPath = [NSFileManager cfm_cacheDirectoryPathForResourceWithPath:@"test/destination"];
    
    [NSFileManager cfm_saveData:self.dataToBeSaved
                         toPath:self.absoluteSourcePath];
}

- (void)tearDown
{
    self.dataToBeSaved = nil;
    self.absolutePath = nil;
    
    if ([NSFileManager cfm_fileExistsAtPath:self.absolutePath])
    {
        [NSFileManager cfm_deleteDataAtPath:self.absolutePath];
    }
    
    if ([NSFileManager cfm_fileExistsAtPath:self.absolutePathWithAdditionalDirectory])
    {
        [NSFileManager cfm_deleteDataAtPath:self.absolutePathWithAdditionalDirectory];
    }
    
    if ([NSFileManager cfm_fileExistsAtPath:self.absolutePathWithoutFile])
    {
        [NSFileManager cfm_deleteDataAtPath:self.absolutePathWithoutFile];
    }
    
    if ([NSFileManager cfm_fileExistsAtPath:self.absoluteDestinationPath])
    {
        [NSFileManager cfm_deleteDataAtPath:self.absoluteDestinationPath];
    }
    
    [super tearDown];
}

#pragma mark - URL

- (void)test_fileURLForPath_url
{
    NSURL *expectedURL = [NSURL fileURLWithPath:self.absolutePath];
    
    NSURL *returnedURL = [NSFileManager cfm_fileURLForPath:self.absolutePath];
    
    XCTAssertEqualObjects(returnedURL, expectedURL, @"URLs returned do not match: %@ and %@", returnedURL, expectedURL);
}

#pragma mark - Saving

- (void)test_saveData_fileOnDisk
{
    [NSFileManager cfm_saveData:self.dataToBeSaved toPath:self.absolutePath];
    
    NSData *dataThatWasSaved = [NSFileManager cfm_retrieveDataAtPath:self.absolutePath];
    
    XCTAssertEqualObjects(self.dataToBeSaved, dataThatWasSaved, @"Data returned do not match: %@ and %@", self.dataToBeSaved, dataThatWasSaved);
}

- (void)test_saveData_successfulSaveReturnValue
{
    BOOL saved = [NSFileManager cfm_saveData:self.dataToBeSaved toPath:self.absolutePath];
    
    XCTAssertTrue(saved, @"Successful save of data should return TRUE");
}

- (void)test_saveData_failedSaveNilDataReturnValue
{
    BOOL saved = [NSFileManager cfm_saveData:nil toPath:self.absolutePath];
    
    XCTAssertFalse(saved, @"Failed save of data should return FALSE");
}

- (void)test_saveData_failedSaveNilPathReturnValue
{
    BOOL saved = [NSFileManager cfm_saveData:self.dataToBeSaved toPath:nil];
    
    XCTAssertFalse(saved, @"Failed save of data should return FALSE");
}

- (void)test_saveData_fileOnDiskWithDirectoryCreation
{
    [NSFileManager cfm_saveData:self.dataToBeSaved toPath:self.absolutePathWithAdditionalDirectory];
    
    NSData *dataThatWasSaved = [NSFileManager cfm_retrieveDataAtPath:self.absolutePathWithAdditionalDirectory];
    
    XCTAssertEqualObjects(self.dataToBeSaved, dataThatWasSaved, @"Data returned do not match: %@ and %@", self.dataToBeSaved, dataThatWasSaved);
}

#pragma mark - Directory

- (void)test_createDirectoryAtPath_directoryCreatedWithFile
{
    [NSFileManager cfm_createDirectoryAtPath:self.absolutePathWithAdditionalDirectory];
    
    BOOL fileExists = [NSFileManager cfm_fileExistsAtPath:self.absolutePathWithAdditionalDirectory];
    
    XCTAssertTrue(fileExists, @"File should exist within a custom directory");
}

- (void)test_createDirectoryAtPath_directoryCreatedWithoutFile
{
    [NSFileManager cfm_createDirectoryAtPath:self.absolutePathWithoutFile];
    
    BOOL fileExists = [NSFileManager cfm_fileExistsAtPath:self.absolutePathWithoutFile];
    
    XCTAssertTrue(fileExists, @"Directory should be created");
}

- (void)test_createDirectoryAtPath_returnValue
{
    BOOL directoryCreated = [NSFileManager cfm_createDirectoryAtPath:self.absolutePathWithAdditionalDirectory];
    
    XCTAssertTrue(directoryCreated, @"TRUE should be returned when custom directory is created");
}

- (void)test_createDirectoryAtPath_failedCreationNilPathReturnValue
{
    BOOL directoryCreated = [NSFileManager cfm_createDirectoryAtPath:nil];
    
    XCTAssertFalse(directoryCreated, @"FALSE should be returned when not path is provided");
}

#pragma mark - FileExists

- (void)test_fileExistsAtPath_trueReturnValue
{
    [NSFileManager cfm_saveData:self.dataToBeSaved toPath:self.absolutePath];
    
    BOOL fileExists = [NSFileManager cfm_fileExistsAtPath:self.absolutePath];
    
    XCTAssertTrue(fileExists, @"File does exist and should be returned as TRUE");
}

- (void)test_fileExistsAtPath_falseReturnValueForFileThatDoesNotExist
{
    NSString *resourceThatDoesNotExist = @"unknown.jpg";
    
    BOOL fileExists = [NSFileManager cfm_fileExistsAtPath:resourceThatDoesNotExist];
    
    XCTAssertFalse(fileExists, @"File does not exist and should be returned as FALSE");
}

- (void)test_fileExistsAtPath_falseReturnValueForNilParameter
{
    BOOL fileExists = [NSFileManager cfm_fileExistsAtPath:nil];
    
    XCTAssertFalse(fileExists, @"Nil parameter and should be returned as FALSE");
}

#pragma mark - FileExistsAsync

- (void)test_fileExistsAtPathAsync_trueReturnValue
{
    [NSFileManager cfm_saveData:self.dataToBeSaved toPath:self.absolutePath];
    
    XCTestExpectation *expectation = [self expectationWithDescription:@"Handler called"];
    
    [NSFileManager cfm_fileExistsAtPath:self.absolutePath
                             completion:^(BOOL fileExists)
     {
         XCTAssertTrue(fileExists, @"File does exist and should be returned as TRUE");
         [expectation fulfill];
     }];
    
    [self waitForExpectationsWithTimeout:1.0
                                 handler:nil];
}

- (void)test_fileExistsAtPathAsync_falseReturnValueForFileThatDoesNotExist
{
    NSString *resourceThatDoesNotExist = @"unknown.jpg";
    
    XCTestExpectation *expectation = [self expectationWithDescription:@"Handler called"];
    
    [NSFileManager cfm_fileExistsAtPath:resourceThatDoesNotExist
                             completion:^(BOOL fileExists)
     {
         XCTAssertFalse(fileExists, @"File does not exist and should be returned as FALSE");
         [expectation fulfill];
     }];
    
    [self waitForExpectationsWithTimeout:1.0
                                 handler:nil];
}

- (void)test_fileExistsAtPathAsync_falseReturnValueForNilParameter
{
    XCTestExpectation *expectation = [self expectationWithDescription:@"Handler called"];
    
    [NSFileManager cfm_fileExistsAtPath:nil
                             completion:^(BOOL fileExists)
     {
         XCTAssertFalse(fileExists, @"Nil parameter and should be returned as FALSE");
         [expectation fulfill];
     }];
    
    [self waitForExpectationsWithTimeout:1.0
                                 handler:nil];
}

- (void)test_fileExistsAtPathAsync_returnedOnCallerThread
{
    [NSFileManager cfm_saveData:self.dataToBeSaved toPath:self.absolutePath];
    
    XCTestExpectation *expectation = [self expectationWithDescription:@"Handler called"];
    
    NSOperationQueue *currentQueue = [NSOperationQueue currentQueue];
    
    [NSFileManager cfm_fileExistsAtPath:self.absolutePath
                             completion:^(BOOL fileExists)
     {
         XCTAssertEqualObjects(currentQueue, [NSOperationQueue currentQueue], @"Should be returned on the celler thread");
         [expectation fulfill];
     }];
    
    [self waitForExpectationsWithTimeout:1.0
                                 handler:nil];

}

#pragma mark - Deletion

- (void)test_deleteDataFromDocumentsDirectoryWithPath_deletesSavedFile
{
    [NSFileManager cfm_saveData:self.dataToBeSaved toPath:self.absolutePath];
    
    [NSFileManager cfm_deleteDataAtPath:self.absolutePath];
    
    BOOL fileExists = [NSFileManager cfm_fileExistsAtPath:self.absolutePath];
    
    XCTAssertFalse(fileExists, @"File should not exist as file should have been deleted, this call should be returned as FALSE");
}

- (void)test_deleteDataFromDocumentsDirectoryWithPath_deletesSavedFileReturnValue
{
    [NSFileManager cfm_saveData:self.dataToBeSaved toPath:self.absolutePath];
    
    BOOL deletion = [NSFileManager cfm_deleteDataAtPath:self.absolutePath];
    
    XCTAssertTrue(deletion, @"File should not exist as file should have been deleted, this call should be returned as FALSE");
}

- (void)test_deleteDataFromDocumentsDirectoryWithPath_falseReturnValueForFileThatDoesNotExist
{
    NSString *resourceThatDoesNotExist = @"unknown.jpg";
    
    BOOL deletion = [NSFileManager cfm_deleteDataAtPath:resourceThatDoesNotExist];
    
    XCTAssertFalse(deletion, @"Deletion should return FALSE as the resource does not exist");
}

- (void)test_deleteDataFromDocumentsDirectoryWithPath_falseReturnValueForNilParameter
{
    BOOL deletion = [NSFileManager cfm_deleteDataAtPath:nil];
    
    XCTAssertFalse(deletion, @"Deletion should return FALSE as the resource is nil");
}

- (void)test_deleteDataFromDocumentsDirectoryWithPath_deletesSavedFileWithFolder
{
    [NSFileManager cfm_saveData:self.dataToBeSaved toPath:self.absolutePathWithAdditionalDirectory];
    
    [NSFileManager cfm_deleteDataAtPath:self.absolutePathWithAdditionalDirectory];
    
    BOOL fileExists = [NSFileManager cfm_deleteDataAtPath:self.absolutePathWithAdditionalDirectory];
    
    XCTAssertFalse(fileExists, @"File should not exist as file should have been deleted, this call should be returned as FALSE");
}

#pragma mark - Move

- (void)test_moveFileFromSourcePath_fileMoved
{
    [NSFileManager cfm_moveFileFromSourcePath:self.absoluteSourcePath
                            toDestinationPath:self.absoluteDestinationPath];
    
    BOOL fileMoved = [NSFileManager cfm_fileExistsAtPath:self.absoluteDestinationPath];
    
    XCTAssertTrue(fileMoved, @"File has not been moved");
}

- (void)test_moveFileFromSourcePath_trueReturnValueWhenFileMoved
{
    BOOL fileMoved = [NSFileManager cfm_moveFileFromSourcePath:self.absoluteSourcePath
                                             toDestinationPath:self.absoluteDestinationPath];
    
    XCTAssertTrue(fileMoved, @"File has not been moved");
}

- (void)test_moveFileFromSourcePath_falseReturnValueWhenNilSource
{
    BOOL fileMoved = [NSFileManager cfm_moveFileFromSourcePath:nil
                                             toDestinationPath:self.absoluteDestinationPath];
    
    XCTAssertFalse(fileMoved, @"Source isn't set so file shouldn't have been moved");
}

- (void)test_moveFileFromSourcePath_falseReturnValueWhenNilDestination
{
    BOOL fileMoved = [NSFileManager cfm_moveFileFromSourcePath:self.absoluteSourcePath
                                             toDestinationPath:nil];
    
    XCTAssertFalse(fileMoved, @"Source isn't set so file shouldn't have been moved");
}

@end

File: ConvenientFileManagerTests/Tests/Persistence/NSFileManagerCFMPersistenceTests.swift
//
//  NSFileManagerCFMPersistenceTests.swift
//  ConvenientFileManager
//
//  Created by William Boles on 29/03/2016.
//  Copyright © 2016 Boles. All rights reserved.
//

import XCTest

class NSFileManagerCFMPersistenceTests: XCTestCase {

    //MARK: Accessor
    
    var dataToBeSaved: NSData = {
        let dataToSaved = "Test string to be converted into data".dataUsingEncoding(NSUTF8StringEncoding)
        
        return dataToSaved!
    }()
    
    var emptyData: NSData = {
        let emptyData = "".dataUsingEncoding(NSUTF8StringEncoding)
        
        return emptyData!
    }()
    
    var testFolder: String = {
        let testFolder = "test"
        
        return testFolder
    }()
    
    lazy var absolutePath: String = {
        let absolutePath = NSFileManager.documentsDirectoryPathForResourceWithPath("test.mp4")
        
        return absolutePath
    }()
    
    lazy var absolutePathWithSpaceInFilename: String = {
        let absolutePathWithSpaceInFilename = NSFileManager.documentsDirectoryPathForResourceWithPath("test 98.mp4")
        
        return absolutePathWithSpaceInFilename
    }()
    
    lazy var absolutePathWithAdditionalDirectory: String = {
        let absolutePathWithAdditionalDirectory = NSFileManager.cacheDirectoryPathForResourceWithPath("\(self.testFolder)/test/test.mp4")
        
        return absolutePathWithAdditionalDirectory
    }()
    
    lazy var absolutePathWithoutFile: String = {
        let absolutePathWithoutFile = NSFileManager.cacheDirectoryPathForResourceWithPath("\(self.testFolder)/testcache")
        
        return absolutePathWithoutFile
    }()
    
    lazy var absoluteSourcePath: String = {
        let absoluteSourcePath = NSFileManager.cacheDirectoryPathForResourceWithPath("\(self.testFolder)/source.png")
        
        return absoluteSourcePath
    }()
    
    lazy var absoluteSourcePathWithSpaceInFilename: String = {
        let absoluteSourcePathWithSpaceInFilename = NSFileManager.cacheDirectoryPathForResourceWithPath("\(self.testFolder)/source 98.png")
        
        return absoluteSourcePathWithSpaceInFilename
    }()
    
    lazy var absoluteDestinationPath: String = {
        let absoluteDestinationPath = NSFileManager.cacheDirectoryPathForResourceWithPath("\(self.testFolder)/destination.png")
        
        return absoluteDestinationPath
    }()
    
    lazy var absoluteDestinationPathWithSpaceInFilename: String = {
        let absoluteDestinationPathWithSpaceInFilename = NSFileManager.cacheDirectoryPathForResourceWithPath("\(self.testFolder)/destination 879.png")
        
        return absoluteDestinationPathWithSpaceInFilename
    }()
    
    //MARK: TestSuiteLifecycle
    
    override func setUp() {
        super.setUp()
        
        NSFileManager.saveData(self.dataToBeSaved, absolutePath: self.absoluteSourcePath)
        NSFileManager.saveData(self.dataToBeSaved, absolutePath: self.absoluteSourcePathWithSpaceInFilename)
    }
    
    override func tearDown() {
        
        if NSFileManager.fileExistsAtPath(self.absolutePath) {
           NSFileManager.deleteDataAtPath(self.absolutePath)
        }
        
        if NSFileManager.fileExistsAtPath(self.absolutePathWithSpaceInFilename) {
            NSFileManager.deleteDataAtPath(self.absolutePathWithSpaceInFilename)
        }
        
        if NSFileManager.fileExistsAtPath(self.absolutePathWithAdditionalDirectory) {
            NSFileManager.deleteDataAtPath(self.absolutePathWithAdditionalDirectory)
        }
        
        if NSFileManager.fileExistsAtPath(self.absolutePathWithoutFile) {
            NSFileManager.deleteDataAtPath(self.absolutePathWithoutFile)
        }
        
        if NSFileManager.fileExistsAtPath(self.absoluteDestinationPath) {
            NSFileManager.deleteDataAtPath(self.absoluteDestinationPath)
        }
        
        if NSFileManager.fileExistsAtPath(self.absoluteSourcePath) {
            NSFileManager.deleteDataAtPath(self.absoluteSourcePath)
        }
        
        if NSFileManager.fileExistsAtPath(self.absoluteDestinationPathWithSpaceInFilename) {
            NSFileManager.deleteDataAtPath(self.absoluteDestinationPathWithSpaceInFilename)
        }
        
        if NSFileManager.fileExistsAtPath(self.absoluteSourcePathWithSpaceInFilename) {
            NSFileManager.deleteDataAtPath(self.absoluteSourcePathWithSpaceInFilename)
        }
        
        let documentsTestFolder = NSFileManager.documentsDirectoryPathForResourceWithPath(self.testFolder)
        
        if NSFileManager.fileExistsAtPath(documentsTestFolder) {
            NSFileManager.deleteDataAtPath(documentsTestFolder)
        }
        
        let cacheTestFolder = NSFileManager.cacheDirectoryPathForResourceWithPath(self.testFolder)
        
        if NSFileManager.fileExistsAtPath(cacheTestFolder) {
            NSFileManager.deleteDataAtPath(cacheTestFolder)
        }
        
        super.tearDown()
    }

    //MARK: URL
    
    func test_fileURLForPath_url() {
        let expectedURL = NSURL(fileURLWithPath: self.absolutePath)
        let returnedURL = NSFileManager.fileURLForPath(self.absolutePath)
        
        XCTAssertEqual(returnedURL, expectedURL, "URLs returned do not match: \(returnedURL) and \(expectedURL)")
    }
    
    //MARK Saving
    
    func test_saveData_fileOnDisk() {
        NSFileManager.saveData(self.dataToBeSaved, absolutePath: self.absolutePath)
        
        let dataThatWasSaved = NSFileManager.retrieveDataAtPath(self.absolutePath)
        
        XCTAssertEqual(self.dataToBeSaved, dataThatWasSaved, "Data returned do not match: \(self.dataToBeSaved) and \(dataThatWasSaved)")
    }
    
    func test_saveData_fileOnDiskWithSpaceInFilename() {
        NSFileManager.saveData(self.dataToBeSaved, absolutePath: self.absolutePathWithSpaceInFilename)
        
        let dataThatWasSaved = NSFileManager.retrieveDataAtPath(self.absolutePathWithSpaceInFilename)
        
        XCTAssertEqual(self.dataToBeSaved, dataThatWasSaved, "Data returned do not match: \(self.dataToBeSaved) and \(dataThatWasSaved)")
    }
    
    func test_saveData_successfulSaveReturnValue() {
        let saved = NSFileManager.saveData(self.dataToBeSaved, absolutePath: self.absolutePath)
        
        XCTAssertTrue(saved, "Successful save of data should return TRUE")
    }
    
    func test_saveData_failedSaveEmptyDataReturnValue() {
        let saved = NSFileManager.saveData(self.emptyData, absolutePath: self.absolutePath)
        
        XCTAssertFalse(saved, "Failed save of data should return FALSE")
    }

    func test_saveData_failedSaveEmptyPathReturnValue() {
        let saved = NSFileManager.saveData(self.dataToBeSaved, absolutePath: "")
        
        XCTAssertFalse(saved, "Failed save of data should return FALSE")
    }
    
    func test_saveData_fileOnDiskWithDirectoryCreation() {
        NSFileManager.saveData(self.dataToBeSaved, absolutePath: self.absolutePathWithAdditionalDirectory)
        
        let dataThatWasSaved = NSFileManager.retrieveDataAtPath(self.absolutePathWithAdditionalDirectory)
        
        XCTAssertEqual(self.dataToBeSaved, dataThatWasSaved, "Data returned do not match: \(self.dataToBeSaved) and \(dataThatWasSaved)")
    }
    
    //MARK: Directory
    
    func test_createDirectoryAtPath_directoryCreatedWithFile() {
        NSFileManager.createDirectoryAtPath(self.absolutePathWithAdditionalDirectory)
        
        let fileExists = NSFileManager.fileExistsAtPath(self.absolutePathWithAdditionalDirectory)
        
        XCTAssertTrue(fileExists, "File should exist within a custom directory")
    }
    
    func test_createDirectoryAtPath_directoryCreatedWithoutFile() {
        NSFileManager.createDirectoryAtPath(self.absolutePathWithoutFile)
        
        let fileExists = NSFileManager.fileExistsAtPath(self.absolutePathWithoutFile)
        
        XCTAssertTrue(fileExists, "Directory should be created")
    }
    
    func test_createDirectoryAtPath_returnValue() {
        let directoryCreated = NSFileManager.createDirectoryAtPath(self.absolutePathWithAdditionalDirectory)
        
        XCTAssertTrue(directoryCreated, "TRUE should be returned when custom directory is created")
    }
    
    func test_createDirectoryAtPath_failedCreationNilPathReturnValue() {
        let directoryCreated = NSFileManager.createDirectoryAtPath("")
        
        XCTAssertFalse(directoryCreated, "FALSE should be returned when not path is provided")
    }

    //MARK: FileExists
    
    func test_fileExistsAtPath_trueReturnValue() {
        NSFileManager.saveData(self.dataToBeSaved, absolutePath: self.absolutePath)
        
        let fileExists = NSFileManager.fileExistsAtPath(self.absolutePath)
        
        XCTAssertTrue(fileExists, "File does exist and should be returned as TRUE")
    }
    
    func test_fileExistsAtPath_trueReturnValueWithSpaceInFilename() {
        NSFileManager.saveData(self.dataToBeSaved, absolutePath: self.absolutePathWithoutFile)
        
        let fileExists = NSFileManager.fileExistsAtPath(self.absolutePathWithoutFile)
        
        XCTAssertTrue(fileExists, "File does exist and should be returned as TRUE")
    }

    func test_fileExistsAtPath_falseReturnValueForFileThatDoesNotExist() {
        let resourceThatDoesNotExist = "unknown.jpg"
        
        let fileExists = NSFileManager.fileExistsAtPath(resourceThatDoesNotExist)
        
        XCTAssertFalse(fileExists, "File does not exist and should be returned as FALSE")
    }
    
    func test_fileExistsAtPath_falseReturnValueForEmptyParameter() {
        let fileExists = NSFileManager.fileExistsAtPath("")
        
        XCTAssertFalse(fileExists, "Empty parameter and should be returned as FALSE")
    }
    
    //MARK FileExistsAsync
    
    func test_fileExistsAtPathAsync_trueReturnValue() {
        NSFileManager.saveData(self.dataToBeSaved, absolutePath: self.absolutePath)
        
        let expectation = self.expectationWithDescription("Handler called")
        
        NSFileManager.fileExistsAtPath(self.absolutePath) { (fileExists) in
            XCTAssertTrue(fileExists, "File does exist and should be returned as TRUE")
            expectation.fulfill()
        }
        
        self.waitForExpectationsWithTimeout(1.0, handler: nil)
    }

    func test_fileExistsAtPathAsync_falseReturnValueForFileThatDoesNotExist() {
        let resourceThatDoesNotExist = "unknown.jpg"
        
        let expectation = self.expectationWithDescription("Handler called")
        
        NSFileManager.fileExistsAtPath(resourceThatDoesNotExist) { (fileExists) in
            XCTAssertFalse(fileExists, "File does not exist and should be returned as FALSE")
            expectation.fulfill()
        }
        
        self.waitForExpectationsWithTimeout(1.0, handler: nil)
    }
    
    func test_fileExistsAtPathAsync_falseReturnValueForEmptyParameter() {
        let expectation = self.expectationWithDescription("Handler called")
        
        NSFileManager.fileExistsAtPath("") { (fileExists) in
            XCTAssertFalse(fileExists, "Empty parameter and should be returned as FALSE")
            expectation.fulfill()
        }
        
        self.waitForExpectationsWithTimeout(1.0, handler: nil)
    }
    
    func test_fileExistsAtPathAsync_returnedOnCallerThread() {
        NSFileManager.saveData(self.dataToBeSaved, absolutePath: self.absolutePath)
        
        let expectation = self.expectationWithDescription("Handler called")
        
        let currentQueue = NSOperationQueue.currentQueue()
        
        NSFileManager.fileExistsAtPath(self.absolutePath) { (fileExists) in
            XCTAssertEqual(currentQueue, NSOperationQueue.currentQueue(), "Should be returned on the celler thread")
            expectation.fulfill()
        }
        
        self.waitForExpectationsWithTimeout(1.0, handler: nil)
    }
    
    //MARK: Deletion
    
    func test_deleteDataFromDocumentsDirectoryWithPath_deletesSavedFile() {
        NSFileManager.saveData(self.dataToBeSaved, absolutePath: self.absolutePath)
        
        NSFileManager.deleteDataAtPath(self.absolutePath)
        
        let fileExists = NSFileManager.fileExistsAtPath(self.absolutePath)
        
        XCTAssertFalse(fileExists, "File should not exist as file should have been deleted, this call should be returned as FALSE")
    }
    
    func test_deleteDataFromDocumentsDirectoryWithPath_deletesSavedFileReturnValue() {
        NSFileManager.saveData(self.dataToBeSaved, absolutePath: self.absolutePath)
        
        let deleted = NSFileManager.deleteDataAtPath(self.absolutePath)
        
        XCTAssertTrue(deleted, "File should not exist as file should have been deleted, this call should be returned as FALSE")
    }
    
    func test_deleteDataFromDocumentsDirectoryWithPath_falseReturnValueForFileThatDoesNotExist() {
        let resourceThatDoesNotExist = "unknown.jpg"
        
        let deleted = NSFileManager.deleteDataAtPath(resourceThatDoesNotExist)
        
        XCTAssertFalse(deleted, "Deletion should return FALSE as the resource does not exist")
    }
    
    func test_deleteDataFromDocumentsDirectoryWithPath_falseReturnValueForEmptyParameter() {
        let deleted = NSFileManager.deleteDataAtPath("")
        
        XCTAssertFalse(deleted, "Deletion should return FALSE as the resource is nil")
    }
    
    func test_deleteDataFromDocumentsDirectoryWithPath_deletesSavedFileWithFolder() {
        NSFileManager.saveData(self.dataToBeSaved, absolutePath: self.absolutePathWithAdditionalDirectory)
        
        NSFileManager.deleteDataAtPath(self.absolutePathWithAdditionalDirectory)
        
        let fileExists = NSFileManager.deleteDataAtPath(self.absolutePathWithAdditionalDirectory)
        
        XCTAssertFalse(fileExists, "File should not exist as file should have been deleted, this call should be returned as FALSE")
    }
    
    //MARK: Move
    
    func test_moveFileFromSourcePath_fileMoved() {
        NSFileManager.moveFile(self.absoluteSourcePath, destinationAbsolutePath: self.absoluteDestinationPath)
        
        let fileMoved = NSFileManager.fileExistsAtPath(self.absoluteDestinationPath)
        
        XCTAssertTrue(fileMoved, "File has not been moved")
    }
    
    func test_moveFileFromSourcePath_fileMovedWithSpaceInSourcePathFileName() {
        NSFileManager.moveFile(self.absoluteSourcePathWithSpaceInFilename, destinationAbsolutePath: self.absoluteDestinationPath)
        
        let fileMoved = NSFileManager.fileExistsAtPath(self.absoluteDestinationPath)
        
        XCTAssertTrue(fileMoved, "File has not been moved")
    }
    
    func test_moveFileFromSourcePath_fileMovedWithSpaceInDestinationPathFileName() {
        NSFileManager.moveFile(self.absoluteSourcePath, destinationAbsolutePath: self.absoluteDestinationPathWithSpaceInFilename)
        
        let fileMoved = NSFileManager.fileExistsAtPath(self.absoluteDestinationPathWithSpaceInFilename)
        
        XCTAssertTrue(fileMoved, "File has not been moved")
    }
    
    func test_moveFileFromSourcePath_trueReturnValueWhenFileMoved() {
        let fileMoved = NSFileManager.moveFile(self.absoluteSourcePath, destinationAbsolutePath: self.absoluteDestinationPath)
        
        XCTAssertTrue(fileMoved, "File has not been moved")
    }

    func test_moveFileFromSourcePath_falseReturnValueWhenEmptySource() {
        let fileMoved = NSFileManager.moveFile("", destinationAbsolutePath: self.absoluteDestinationPath)
        
        XCTAssertFalse(fileMoved, "Source is empty so file shouldn't have been moved")
    }
    
    func test_moveFileFromSourcePath_falseReturnValueWhenEmptyDestination() {
        let fileMoved = NSFileManager.moveFile(self.absoluteSourcePath, destinationAbsolutePath: "")
        
        XCTAssertFalse(fileMoved, "Destination is empty so file shouldn't have been moved")
    }
 
}
